RecoLocalCalo/EcalRecAlgos/src/EcalUncalibRecHitMultiFitAlgo.cc

```diff
@@ -20,70 +25,142 @@ EcalUncalibRecHitMultiFitAlgo::EcalUncalibRecHitMultiFitAlgo() :
+      gainsPedestal[iSample] = dynamicPedestal ? 1 : -1; //-1 for static pedestals
     }
 
-    amplitude = ((double)(sample.adc()) - pedestal) * gainratio;
+    if (dynamicPedestal) {
+      amplitude = (double)(sample.adc())*gainratio;
+    }
+    else {
+      amplitude = ((double)(sample.adc()) - pedestal) * gainratio;
+    }
     
     if (gainId == 0) {
+       edm::LogError("EcalUncalibRecHitMultiFitAlgo")<< "Saturation encountered.  Multifit is not intended to be used for saturated channels.";
       //saturation
-      amplitude = (4095. - pedestal) * gainratio;
+      if (dynamicPedestal) {
+        amplitude = 4095.*gainratio;
+      }
+      else {
+        amplitude = (4095. - pedestal) * gainratio;
+      }
     }
         
     amplitudes[iSample] = amplitude;
     
-    if (amplitude>maxamplitude) {
+    if (gainratio > maxgainratio || gainId==0 || amplitude>maxamplitude) {
     //if (iSample==5) {
       maxamplitude = amplitude;
+      maxgainratio = gainratio;
+      iSampleMax = iSample;
       pedval = pedestal;
-      pedrms = pederr*gainratio;
-    }    
+    }
         
   }
   
+  //special handling for gain switch, where sample before maximum is potentially affected by slew rate limitation
+  //optionally apply a stricter criteria, assuming slew rate limit is only reached in case where maximum sample has gain switched but previous sample has not
+  //A floating negative single-sample offset is added to the fit
+  //such that the affected sample is treated only as a lower limit for the true amplitude
+  bool mitigateBadSample = _mitigateBadSamples && hasGainSwitch && iSampleMax>0;
+  mitigateBadSample &= (!_selectiveBadSampleCriteria || (gainsNoise.coeff(iSampleMax-1)!=gainsNoise.coeff(iSampleMax)) );
+  if (mitigateBadSample) {
+    badSamples[iSampleMax-1] = 1;
+  }
+  
   double amplitude, amperr, chisq;
   bool status = false;
   
+  //compute noise covariance matrix, which depends on the sample gains
+  SampleMatrix noisecov;
+  if (hasGainSwitch) {
+    std::array<double,3> pedrmss = {{aped->rms_x12, aped->rms_x6, aped->rms_x1}};
+    std::array<double,3> gainratios = {{ 1., aGain->gain12Over6(), aGain->gain6Over1()*aGain->gain12Over6()}};
+    if (_simplifiedNoiseModelForGainSwitch) {
+      int gainidxmax = gainsNoise[iSampleMax];
+      noisecov = gainratios[gainidxmax]*gainratios[gainidxmax]*pedrmss[gainidxmax]*pedrmss[gainidxmax]*noisecors[gainidxmax];
+      if (!dynamicPedestal && _addPedestalUncertainty>0.) {
+        //add fully correlated component to noise covariance to inflate pedestal uncertainty
+        noisecov += _addPedestalUncertainty*_addPedestalUncertainty*SampleMatrix::Ones();
+      }
+    }
+    else {
+      noisecov = SampleMatrix::Zero();
+      for (unsigned int gainidx=0; gainidx<noisecors.size(); ++gainidx) {
+        SampleGainVector mask = gainidx*SampleGainVector::Ones();
+        SampleVector pedestal = (gainsNoise.array()==mask.array()).cast<SampleVector::value_type>();
```

Comment:
It doesn't show up in the documentation because it's aliased in a strange way with macros inside the code, but it's pointing back to http://eigen.tuxfamily.org/dox/classEigen_1_1ArrayBase.html#a591a8024b2762dfdc4947cdcb9f74fdd

So it's an element wise comparison and returns a vector of bools (which is why the cast is needed)